dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+    { IDS_COL_SIZE, LVCFMT_RIGHT, 16, SHCOLSTATE_TYPE_INT | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_SIZE },
+    { IDS_COL_TYPE, LVCFMT_LEFT, 20, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_STORAGETYPE },
+    { IDS_COL_MDATE, LVCFMT_LEFT, 20, SHCOLSTATE_TYPE_DATE | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_WRITETIME },
+    { IDS_COL_PATH, LVCFMT_LEFT, 30, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, &CLSID_CabFolder, 0 },
+    { IDS_COL_ATT, LVCFMT_RIGHT, 10, SHCOLSTATE_TYPE_STR, &FMTID_Storage, PID_STG_ATTRIBUTES },
+};
+
+static CABITEM* CreateItem(LPCWSTR Path, UINT Attrib, UINT Size, UINT DateTime)
+{
+    const SIZE_T len = lstrlenW(Path), cb = FIELD_OFFSET(CABITEM, Path[len + 1]);
+    if (cb > 0xffff)
+        return NULL;
+    CABITEM *p = (CABITEM*)SHAlloc(cb + sizeof(USHORT));
+    if (p)
+    {
+        p->cb = (USHORT)cb;
+        p->Unknown = 0;
+        p->Size = Size;
+        p->Attrib = Attrib;
+        p->Date = HIWORD(DateTime);
+        p->Time = LOWORD(DateTime);
+        p->NameOffset = 0;
+        for (UINT i = 0;; ++i)
+        {
+            WCHAR c = Path[i];
+            if (c == L':') // Don't allow absolute paths
+                c = L'_';
+            if (c == L'/') // Normalize
+                c = L'\\';
+            if (c == '\\')
+                p->NameOffset = i + 1;
+            p->Path[i] = c;
+            if (!c)
+                break;
+        }
+        ((SHITEMID*)((BYTE*)p + cb))->cb = 0;
+    }
+    return p;
+}
+
+static CABITEM* CreateItem(LPCSTR Path, UINT Attrib, UINT Size = 0, UINT DateTime = 0)
+{
+    WCHAR buf[MAX_PATH * 2];
+    UINT codepage = (Attrib & _A_NAME_IS_UTF) ? CP_UTF8 : CP_ACP;
+    if (MultiByteToWideChar(codepage, 0, Path, -1, buf, _countof(buf)))
+        return CreateItem(buf, Attrib, Size, DateTime);
+    return NULL;
+}
+
+static HRESULT CALLBACK ItemMenuCallback(IShellFolder *psf, HWND hwnd, IDataObject *pdtobj,
+                                         UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    enum { IDC_EXTRACT, IDC_EXTRACTALL };
+    HRESULT hr = E_NOTIMPL;
+    const BOOL Background = !pdtobj;
+
+    switch (uMsg)
+    {
+        case DFM_MODIFYQCMFLAGS:
+        {
+            *((UINT*)lParam) = wParam | CMF_NOVERBS | (Background ? 0 : CMF_VERBSONLY);
+            return S_OK;
+        }
+
+        case DFM_MERGECONTEXTMENU:
+        {
+            QCMINFO &qcmi = *(QCMINFO*)lParam;
+            UINT pos = qcmi.indexMenu, id = 0;
+            if (Background || SUCCEEDED(hr = InsertMenuItem(qcmi, pos, id, IDC_EXTRACT, IDS_EXTRACT, MFS_DEFAULT)))
+            {
+                hr = InsertMenuItem(qcmi, pos, id, IDC_EXTRACTALL, IDS_EXTRACTALL);
+                if (SUCCEEDED(hr) && !Background)
+                {
+                    --pos;
+                    InsertMenuItem(qcmi, pos, id, 0, -1); // Separator
```

Comment:
Should the extension itself handle this, or is this the job for the host?
(It wouldn't surprise me if this would only be needed on ReactOS tbh)